### PR TITLE
Additional validation of receipts

### PR DESCRIPTION
**Pull request by clokep**
This improves the error messages sent to clients if they send bogus receipts and would have made my life much easier when I was debugging something.

For federation the best we do is check that the server is in the room and drop it if not (this is similar to what we do for typing, see #13830.

This essentially expands the validation added in #14632, and somewhat backs out https://github.com/matrix-org/synapse/commit/9f797a24a452a513628263b1b03172cee20a9856, but I still don't understand how that could ever happen. (Note that instead of raising an error if the event is unknown, we instead ignore the receipt. This should also help in situations where a room is purged, kind of like #15464.)

Sorry for the large diff -- one of the commits in the middle moves some tests around.

### CODE REVIEW

File: tests/rest/client/test_receipts.py [hunk 1]
```diff
@@ -24,30 +29,113 @@
...
+        """Receipts sent for unknown events are ignored."""
+        # Attempt to send a receipt to an unknown room.
         channel = self.make_request(
             "POST",
             "/rooms/!abc:beep/receipt/m.read/$def",
             content={},
-            access_token=self.owner_tok,
+            access_token=self.tok2,
+        )
+        self.assertEqual(channel.code, 200, channel.result)
```
Comment (DMRobertson) [inline tests/rest/client/test_receipts.py:79]: I was surprised that this returns 200 (as does the call below). Talking to Patrick, this is to support message retention.

File: tests/rest/client/test_receipts.py [hunk 2]
```diff
@@ -71,6 +159,123 @@ def test_send_receipt_invalid_receipt_type(self) -> None:
...
+        channel = self.make_request(
+            "POST",
+            f"/rooms/{self.room_id}/receipt/{ReceiptTypes.READ_PRIVATE}/{res['event_id']}",
+            {},
+            access_token=self.tok2,
+        )
+        self.assertEqual(channel.code, 200)
+
+        # Test that we didn't override the public read receipt
+        self.assertIsNone(self._get_read_receipt())
```
Comment (DMRobertson) [inline tests/rest/client/test_receipts.py:240]: Err, is this correct? On 288 above this function returns not-None. I'd expect that to still be the case here, if we haven't overridden anything.

Comment (clokep) [reply tests/rest/client/test_receipts.py:240]: FTR I simply moved this test over, it isn't new.

Comment (DMRobertson) [reply tests/rest/client/test_receipts.py:240]: Ohhhhh. This fetches a receipt from /sync. So it returning `None` means there are no new receipts to tell the client about.

Comment (clokep) [reply tests/rest/client/test_receipts.py:240]: Yes, pretty much it is testing that you can't make something private after you've made it public.

File: tests/rest/client/test_receipts.py [hunk 3]
```diff
@@ -24,30 +29,113 @@
...
+            "POST",
+            f"/rooms/{self.room_id}/receipt/{ReceiptTypes.READ}/{res['event_id']}",
+            {},
+            access_token=self.tok2,
+        )
+        self.assertEqual(channel.code, 200)
+        self.assertNotEqual(self._get_read_receipt(), None)
+
+    def test_send_receipt_unknown_event(self) -> None:
+        """Receipts sent for unknown events are ignored."""
```
Comment (clokep) [inline tests/rest/client/test_receipts.py:71]:
```suggestion
        """Receipts sent for unknown events are ignored to not break message retention."""
```

**Pull request merged**